File: python/src/main/python/Charizard.py
from PokemonCharacter import *
from QuickAttackBehavior import *
from SmallFireBehavior import *


class Charizard(PokemonCharacter):
    def __init__(self):
        self.HIT_POINTS = 77
        self.DEFENSE_MULTIPLIER = 0.3
        self.QuickAttackBehavior = QuickAttackBehavior()
        self.SmallFireBehavior = SmallFireBehavior()
        self.setType("fire")
        self.setName("Charizard")
        self.setSecondaryAttackBehavior(self.QuickAttackBehavior)
        self.setMainAttackBehavior(self.SmallFireBehavior)
        self.setHasEvolution(True)
        self.setHitPoints(self.HIT_POINTS)
        self.setDefenseMultiplier(self.DEFENSE_MULTIPLIER)

    def evolve(self):
        return None

    def defend(self, attack):
        damage = (attack * self.getDefenseMultiplier())
        newHP = self.getHitPoints() - damage
        self.setHitPoints(newHP)
        string = "Defending attack, damage caused is "
        string += str(int(damage))
        string += " new HP is "
        string += str(int(newHP))
        return string

    def secondAttack(self):
        return self.getSecondAttackBehavior().attack()

    def mainAttack(self):
        return self.getMainAttackBehavior().attack()

    def setNewAttack(self, attack, newAttack):
        if (attack == 1):
            self.setMainAttack(newAttack)
        else:
            self.setSecondAttack(newAttack)

File: java/src/main/java/com/iteso/nintendo/Charizard.java
package com.iteso.nintendo;

/**
 * Created by rvillalobos on 2/24/18.
 */
public class Charizard extends PokemonCharacter {
    /**
     * Hit Points are the maximum life of pokemon.
     */
    public static final int HIT_POINTS = 100;
    /**
     * Defense multiplier value between 0-1.
     */
    public static final double DEFENSE_MULTIPLIER = 0.7;
    /**
     * Damage from 1 - 20.
     */
    public static final int MAIN_ATTACK_DAMAGE = 6;
    /**
     * Damage from 1 -25.
     */
    public static final int SECOND_ATTACK_DAMAGE = 22;

    /**
     * Charizard constructor.
     */
    public Charizard() {
        setType("fire");
        setName("Charizard");
        setHasEvolution(true);
        Attack mainAttack = new Tackle();
        Attack secondAttack = new Hadouken();
        setSecondAttack(secondAttack);
        setMainAttack(mainAttack);
        setHitPoints(HIT_POINTS);
        setDefenseMultiplier(DEFENSE_MULTIPLIER);
        mainAttack.setAttackDamage(MAIN_ATTACK_DAMAGE);
        secondAttack.setAttackDamage(SECOND_ATTACK_DAMAGE);
    }

    @Override
    public final String evolve() {
        return null;
    }

    @Override
    public final String defend(final int attack) {
        int damage;

        damage = (int) (attack * getDefenseMultiplier());
        int newHP = getHitPoints() - damage;

        String defendMessage = new String(
                "Defending attack, damage caused is "
                + damage
                + " new HP is "
                + newHP);

        setHitPoints(newHP);
        return defendMessage;

    }

    @Override
    public final String secondAttack() {

        String attackMessage = new String(
                "Attacking opponent with "
                        + getSecondAttack().attack()
                        + " causing a damage of "
                        + getSecondAttack().getAttackDamage());
        return attackMessage;

    }

    @Override
    public final String mainAttack() {
        String attackMessage = new String(
                "Attacking opponent with "
                        + getMainAttack().attack()
                        + " causing a damage of "
                        + getMainAttack().getAttackDamage());
        return attackMessage;

    }

    @Override
    public final void setNewAttack(final int attack,
                                   final int attackDamage,
                                   final Attack newAttack) {
        if (attack == 1) {
            setMainAttack(newAttack);
            getMainAttack().setAttackDamage(attackDamage);
        } else {
            setSecondAttack(newAttack);
            getSecondAttack().setAttackDamage(attackDamage);
        }
    }

}
